cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -225,74 +296,124 @@ protected void run() {
-          continue;
-        }
+    return latestPublishTime;
+  }
+
+  private void updateConsumeRequests(Map<String, List<MonitorMessage>> monitorResponses) {
+    for (Map.Entry<String, List<MonitorMessage>> monitorResponse : monitorResponses.entrySet()) {
+      String topicConfig = monitorResponse.getKey();
+      List<MonitorMessage> monitorMessages = monitorResponse.getValue();
+
+      if (monitorMessages.isEmpty()) {
+        continue;
       }
 
-      String topic = requestKeyToLocalTopic.get(topicConfig);
-      latestPublishTime = Math.max(publish(topicConfig, topic, monitorMessages), latestPublishTime);
+      updateTopicToRequest(topicConfig, monitorMessages);
     }
-    return latestPublishTime;
+  }
+
+  private long updateProgramFinishTime(Map<String, List<MonitorMessage>> monitorResponses) {
+    List<MonitorMessage> programStatusMessages = monitorResponses.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
+    long finishTime = programFinishTime;
+
+    if (finishTime >= 0 || programStatusMessages.isEmpty()) {
+      return finishTime;
+    }
+
+    finishTime = findProgramFinishTime(programStatusMessages);
+
+    if (finishTime >= 0) {
+      // Buffer the program state messages and don't publish them until the end
+      // Otherwise, once we publish, the deprovisioner will kick in and delete the cluster
+      // which could result in losing the last set of messages for some topics.
+      lastProgramStateMessages.addAll(programStatusMessages);
+      // We still update the in memory store for the next fetch offset to avoid fetching duplicate
+      // messages, however, that shouldn't be persisted to avoid potential loss of messages
+      // in case of failure
+      updateTopicToRequest(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC, programStatusMessages);
+    }
+
+    return finishTime;
   }
 
   /**
-   * Publish the given set of messages to the local TMS.
+   * Transactionally publish the given set of messages to the local TMS.
    *
    * @return the latest message publish time or {@code -1L} if there is no message to process
    */
-  private long publish(String topicConfig, String topic, List<MonitorMessage> messages) throws Exception {
-    // TODO publish messages transactionally along with offset table updates
+  private long publish(String topicConfig, String topic, List<MonitorMessage> messages,
+                       AppMetadataStore store) throws Exception {
+    // publish messages to tms
+    MessagePublisher messagePublisher = messagingContext.getMessagePublisher();
     messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), topic,
                              messages.stream().map(s -> s.getMessage().getBytes(StandardCharsets.UTF_8)).iterator());
 
-    MonitorMessage lastMessage = updateTopicToRequest(topicConfig, messages);
+    // persist the last published message as offset in meta store
+    MonitorMessage lastMessage = messages.get(messages.size() - 1);
```

Comment:
Better use `Deque` instead of `List` as the type when deserializing so that you can simply call `getLast` to get the last element.